test/core/security/grpc_authorization_engine_test.cc

```diff
@@ -16,12 +16,106 @@
 
 #include "src/core/lib/security/authorization/grpc_authorization_engine.h"
 
+#include <memory>
+
 #include <gmock/gmock.h>
 #include <gtest/gtest.h>
 
+#include <grpc/grpc_audit_logging.h>
+#include <grpc/grpc_security_constants.h>
+
+#include "src/core/lib/security/authorization/audit_logging.h"
+#include "test/core/util/evaluate_args_test_util.h"
+
 namespace grpc_core {
 
-TEST(GrpcAuthorizationEngineTest, AllowEngineWithMatchingPolicy) {
+namespace {
+
+constexpr absl::string_view kLoggerName = "test_logger";
+constexpr absl::string_view kPolicyName = "authz";
+constexpr absl::string_view kSpiffeId = "spiffe://foo";
+constexpr absl::string_view kRpcMethod = "/foo.Bar/Echo";
+
+using experimental::AuditContext;
+using experimental::AuditLogger;
+using experimental::AuditLoggerFactory;
+using experimental::AuditLoggerRegistry;
+using experimental::RegisterAuditLoggerFactory;
+
+// This test class copies the audit context.
+struct TestAuditContext {
+  explicit TestAuditContext(const AuditContext& context)
+      : rpc_method(context.rpc_method()),
+        principal(context.principal()),
+        policy_name(context.policy_name()),
+        matched_rule(context.matched_rule()),
+        authorized(context.authorized()) {}
+
+  std::string rpc_method;
+  std::string principal;
+  std::string policy_name;
+  std::string matched_rule;
+  bool authorized;
+};
+
+class TestAuditLogger : public AuditLogger {
+ public:
+  explicit TestAuditLogger(
+      std::vector<std::unique_ptr<TestAuditContext>>& contexts)
+      : contexts_(contexts) {}
+
+  void Log(const AuditContext& context) override {
+    contexts_.push_back(std::make_unique<TestAuditContext>(context));
+  }
+
+ private:
+  std::vector<std::unique_ptr<TestAuditContext>>& contexts_;
+};
+
+class TestAuditLoggerFactory : public AuditLoggerFactory {
+ public:
+  explicit TestAuditLoggerFactory(
+      std::vector<std::unique_ptr<TestAuditContext>>& contexts)
+      : contexts_(contexts) {}
+
+  class TestAuditLoggerConfig : public AuditLoggerFactory::Config {
```

Comment:
As per https://google.github.io/styleguide/cppguide.html#Declaration_Order, nested classes should be declated before the ctor.

Reply:
Fixed. Thanks for the tip!